File: src/extractors/sites/pattern_extractor.cpp
#include "pattern_extractor.hpp"
#include "../../http/http_client.hpp"
#include "../../auth/cookie_manager.hpp"
#include "../../paths.hpp"

#include <fstream>
#include <iostream>
#include <set>
#include <nlohmann/json.hpp>

using json =
    nlohmann::json;

PatternExtractor::PatternExtractor()
{
    load_rules(
        "patterns.json"
    );
}

void PatternExtractor::load_rules(
    const std::string& path
)
{
    std::ifstream file(
        Paths::resolve(path)
    );

    if (!file.is_open())
    {
        return;
    }

    json config;

    try
    {
        config =
            json::parse(file);
    }
    catch (const std::exception& e)
    {
        std::cerr
            << "PatternExtractor: JSON parse error: "
            << e.what()
            << "\n";

        return;
    }

    if (!config.is_array())
    {
        std::cerr
            << "PatternExtractor: expected JSON array\n";

        return;
    }

    for (const auto& entry : config)
    {
        PatternRule rule;

        if (!entry.contains("name"))
        {
            continue;
        }

        rule.name =
            entry["name"];

        if (entry.contains("url_match"))
        {
            try
            {
                rule.url_match =
                    std::regex(
                        entry["url_match"]
                    );
            }
            catch (const std::regex_error& e)
            {
                std::cerr
                    << "PatternExtractor: invalid regex for "
                    << rule.name
                    << ": "
                    << e.what()
                    << "\n";

                continue;
            }
        }
        else
        {
            continue;
        }

        if (entry.contains("media_patterns"))
        {
            for (const auto& pattern : entry["media_patterns"])
            {
                try
                {
                    rule.media_patterns.push_back(
                        std::regex(
                            pattern
                        )
                    );
                }
                catch (const std::regex_error& e)
                {
                    std::cerr
                        << "PatternExtractor: invalid media regex for "
                        << rule.name
                        << ": "
                        << e.what()
                        << "\n";
                }
            }
        }

        if (entry.contains("headers"))
        {
            for (const auto& [key, value] : entry["headers"].items())
            {
                rule.headers[key] =
                    value;
            }
        }

        if (entry.contains("cookies_from"))
        {
            rule.cookies_from =
                entry["cookies_from"];
        }

        rules.push_back(
            std::move(rule)
        );
    }
}

bool PatternExtractor::matches(
    const std::string& url
)
{
    for (const auto& rule : rules)
    {
        if (
            std::regex_search(
                url,
                rule.url_match
            )
        )
        {
            return true;
        }
    }

    return false;
}

std::string
PatternExtractor::fetch_page(
    const std::string& url,
    const PatternRule& rule
)
{
    HttpClient client;

    HttpRequest request;

    request.url = url;

    if (!rule.cookies_from.empty())
    {
        request.cookies =
            CookieManager::load(
                rule.cookies_from
            );
    }

    for (const auto& [key, value] : rule.headers)
    {
        request.headers.push_back(
            key
            + ": "
            + value
        );
    }

    if (rule.headers.contains("Referer"))
    {
        request.referer =
            rule.headers.at("Referer");
    }

    HttpResponse response =
        client.get(request);

    if (response.status_code >= 400)
    {
        std::cerr
            << "PatternExtractor: HTTP "
            << response.status_code
            << " for "
            << url
            << "\n";

        return "";
    }

    return response.body;
}

std::string
PatternExtractor::resolve_url(
    const std::string& page_url,
    const std::string& media_url
)
{
    if (
        media_url.starts_with("http://")
        ||
        media_url.starts_with("https://")
    )
    {
        return media_url;
    }

    if (
        media_url.starts_with("//")
    )
    {
        return
            "https:"
            + media_url;
    }

    size_t protocol_pos =
        page_url.find("://");

    if (protocol_pos == std::string::npos)
    {
        return media_url;
    }

    size_t domain_start =
        protocol_pos + 3;

    size_t path_pos =
        page_url.find(
            '/',
            domain_start
        );

    std::string base;

    if (path_pos == std::string::npos)
    {
        base = page_url;
    }
    else
    {
        base =
            page_url.substr(
                0,
                path_pos
            );
    }

    if (
        !media_url.empty()
        &&
        media_url.front() == '/'
    )
    {
        return base + media_url;
    }

    return
        base
        + "/"
        + media_url;
}

std::vector<DownloadTask>
PatternExtractor::extract(
    const std::string& url
)
{
    const PatternRule* matched_rule =
        nullptr;

    for (const auto& rule : rules)
    {
        if (
            std::regex_search(
                url,
                rule.url_match
            )
        )
        {
            matched_rule = &rule;

            break;
        }
    }

    if (!matched_rule)
    {
        return {};
    }

    std::string html =
        fetch_page(
            url,
            *matched_rule
        );

    if (html.empty())
    {
        return {};
    }

    std::vector<std::string> media_urls;

    for (const auto& pattern : matched_rule->media_patterns)
    {
        std::sregex_iterator begin(
            html.begin(),
            html.end(),
            pattern
        );

        std::sregex_iterator end;

        for (
            auto it = begin;
            it != end;
            ++it
        )
        {
            media_urls.push_back(
                (*it)[1]
            );
        }
    }

    for (auto& media_url : media_urls)
    {
        media_url =
            resolve_url(
                url,
                media_url
            );
    }

    std::set<std::string> unique(
        media_urls.begin(),
        media_urls.end()
    );

    media_urls.assign(
        unique.begin(),
        unique.end()
    );

    std::vector<DownloadTask> tasks;

    for (const auto& media_url : media_urls)
    {
        DownloadTask task;

        task.url = media_url;

        if (!matched_rule->cookies_from.empty())
        {
            task.cookies =
                CookieManager::load(
                    matched_rule->cookies_from
                );
        }

        for (const auto& [key, value] : matched_rule->headers)
        {
            task.headers.push_back(
                key
                + ": "
                + value
            );
        }

        size_t last_slash =
            media_url.find_last_of('/');

        if (last_slash != std::string::npos)
        {
            task.filename =
                media_url.substr(
                    last_slash + 1
                );
        }
        else
        {
            task.filename =
                "downloaded_file";
        }

        size_t query_pos =
            task.filename.find('?');

        if (query_pos != std::string::npos)
        {
            task.filename =
                task.filename.substr(
                    0,
                    query_pos
                );
        }

        tasks.push_back(task);
    }

    return tasks;
}

File: android/app/src/main/java/com/mediadownloader/extractors/sites/PatternExtractor.kt
package com.mediadownloader.extractors.sites

import com.mediadownloader.auth.CookieManager
import com.mediadownloader.auth.Paths
import com.mediadownloader.extractors.BaseExtractor
import com.mediadownloader.extractors.DownloadTask
import com.mediadownloader.http.HttpClient
import com.mediadownloader.http.HttpRequest
import kotlinx.serialization.json.*

class PatternExtractor : BaseExtractor {
    private val json = Json { ignoreUnknownKeys = true }
    private val rules = mutableListOf<PatternRule>()

    init {
        loadRules("patterns.json")
    }

    data class PatternRule(
        val name: String,
        val urlMatch: Regex,
        val mediaPatterns: List<Regex>,
        val headers: Map<String, String> = emptyMap(),
        val cookiesFrom: String = ""
    )

    override fun matches(url: String): Boolean {
        return rules.any { it.urlMatch.containsMatchIn(url) }
    }

    override suspend fun extract(url: String): List<DownloadTask> {
        val rule = rules.firstOrNull { it.urlMatch.containsMatchIn(url) } ?: return emptyList()

        val html = fetchPage(url, rule)
        if (html.isEmpty()) return emptyList()

        val mediaUrls = mutableSetOf<String>()
        for (pattern in rule.mediaPatterns) {
            for (match in pattern.findAll(html)) {
                mediaUrls.add(match.groupValues[1])
            }
        }

        val cookies = if (rule.cookiesFrom.isNotEmpty()) {
            CookieManager.load(rule.cookiesFrom)
        } else ""

        return mediaUrls.map { mediaUrl ->
            val resolved = resolveUrl(url, mediaUrl)
            val filename = resolved.substringAfterLast('/')
                .substringBefore('?')
                .ifEmpty { "downloaded_file" }

            DownloadTask(
                url = resolved,
                filename = filename,
                headers = rule.headers,
                cookies = cookies
            )
        }
    }

    private suspend fun fetchPage(url: String, rule: PatternRule): String {
        val resp = HttpClient.get(
            HttpRequest(
                url = url,
                userAgent = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36",
                referer = rule.headers["Referer"] ?: "",
                headers = rule.headers,
                cookies = if (rule.cookiesFrom.isNotEmpty()) {
                    CookieManager.load(rule.cookiesFrom)
                } else ""
            )
        )
        return if (resp.statusCode < 400) resp.body else ""
    }

    private fun resolveUrl(pageUrl: String, mediaUrl: String): String {
        if (mediaUrl.startsWith("http://") || mediaUrl.startsWith("https://")) {
            return mediaUrl
        }
        if (mediaUrl.startsWith("//")) {
            return "https:$mediaUrl"
        }

        val protocolPos = pageUrl.indexOf("://")
        if (protocolPos == -1) return mediaUrl

        val domainStart = protocolPos + 3
        val pathPos = pageUrl.indexOf('/', domainStart)
        val base = if (pathPos == -1) pageUrl else pageUrl.substring(0, pathPos)

        return if (mediaUrl.startsWith("/")) "$base$mediaUrl" else "$base/$mediaUrl"
    }

    private fun loadRules(path: String) {
        val resolved = Paths.resolve(path)
        val file = java.io.File(resolved)
        if (!file.exists()) return

        val text = try { file.readText() } catch (_: Exception) { return }

        val config = try {
            json.parseToJsonElement(text).jsonArray
        } catch (_: Exception) { return }

        for (entry in config) {
            val obj = entry.jsonObject
            val name = obj["name"]?.jsonPrimitive?.content ?: continue
            val urlMatchStr = obj["url_match"]?.jsonPrimitive?.content ?: continue
            val urlMatch = try {
                Regex(urlMatchStr)
            } catch (_: Exception) { continue }

            val mediaPatterns = mutableListOf<Regex>()
            for (pattern in obj["media_patterns"]?.jsonArray ?: emptyList()) {
                try {
                    mediaPatterns.add(Regex(pattern.jsonPrimitive.content))
                } catch (_: Exception) {}
            }

            val headers = mutableMapOf<String, String>()
            for ((key, value) in obj["headers"]?.jsonObject ?: emptyMap()) {
                headers[key] = value.jsonPrimitive.content
            }

            val cookiesFrom = obj["cookies_from"]?.jsonPrimitive?.content ?: ""

            rules.add(PatternRule(name, urlMatch, mediaPatterns, headers, cookiesFrom))
        }
    }
}
